compiler/il/OMRNode.cpp

```diff
@@ -267,6 +269,16 @@ OMR::Node::Node(TR::Node * from, uint16_t numChildren)
    if(comp->getDebug())
       comp->getDebug()->newNode(self());
 
+   TR_IlGenerator * ilGen = comp->getCurrentIlGenerator();
+   if (ilGen)
```

Comment:
What is the rationale for this change ? While I understand that you would be allowing more nodes to be profiled with this change (in particular, the ones that are created during IL gen), the intent of the original code may have been to avoid profiling any nodes that are created based on some other node, i.e. not created directly from bytecodes. I guess I'm wondering what the case is that you are going after with this change and whether this change only affects the case you are interested in.

Reply:
`supportsInduceOSR` looks at `doNotProfile` flag on a bci to see if a node is created during ilgen/directly from bytecodes, to determine if OSR is possible at a given bci. The intention of this change is to ensure that the flag is correctly set, such that the `supportsInduceOSR` does not respond true to nodes that are created after ilgen through this API.

Reply:
I wonder if we can initialize `doNotProfile` in `TR_ByteCodeInfo`'s constructor to 1 such that we only need to set it to 0 when the node is created directly from bytecode.

Reply:
I'm still not completely clear. I assume you are concerned about ```doNotProfile``` being set to false in cases when it ought to be set to true, judging by your comment around not wanting ```supportsInduceOSR``` to return true for nodes created after IL gen. Can you give me an example of code that creates a node after IL gen but was setting ```doNotProfile``` to false ?

Reply:
To be clear my prior comment was a response to your statement : "The intention of this change is to ensure that the flag is correctly set, such that the supportsInduceOSR does not respond true to nodes that are created after ilgen through this API."...I was asking about an example that went through the node api but was presumably creating a node after IL gen that had ```doNotProfile``` set to false.

Reply:
A few examples from existing code are
https://github.com/eclipse/omr/blob/master/compiler/optimizer/LoopVersioner.cpp#L5309
https://github.com/eclipse/omr/blob/master/compiler/optimizer/OMRValuePropagation.cpp#L6485
https://github.com/eclipse/omr/blob/master/compiler/optimizer/OMRValuePropagation.cpp#L7958
https://github.com/eclipse/omr/blob/master/compiler/optimizer/VPHandlers.cpp#L4193
https://github.com/eclipse/omr/blob/master/compiler/optimizer/LoopVersioner.cpp#L5176

They are a lot of places like above where a node is created without passing an originating bytecode node. `doNotProfile` flag on their bci is not guaranteed to be set.

Reply:
I believe if you create a new node with at least one child, the bytecode used is from the first child. If a node is being created without any children, then an originating bytecode node has to be passed in explicitly (and I don't know of cases when it's not). So in a sense, whatever the node you are creating, you do have an originating bytecode node available that ought to be used to call the node api that should be setting ```doNotProfile``` to true for the node being created.

Reply:
From your list of links, I think:

https://github.com/eclipse/omr/blob/master/compiler/optimizer/OMRValuePropagation.cpp#L7958
https://github.com/eclipse/omr/blob/master/compiler/optimizer/VPHandlers.cpp#L4193
https://github.com/eclipse/omr/blob/master/compiler/optimizer/LoopVersioner.cpp#L5176

are concerning since we do see examples of a node without a child being created without an originating bytecode node being passed in. I am not sure what the bytecode info would be for such nodes (you can check in the log file for the compilation). In my opinion these cases ought to be fixed...maybe separately from this PR but it is kind of related to the work done in this PR for handling the problematic cases when originating node is not available.

Reply:
I've checked that `doNotProfile` is only set to `true` when an originating bytecode node is passed, and for branch nodes. This is why I have changes to set `doNotProfile` in Node APIs that don't take an originating bytecode node. One solution for `doNotProfile` is to initialize it to `true` in `TR_ByteCodeInfo`'s constructor and allow Node APIs to set it to false.  Only nodes created in ilgen need it to be set to `false`.

For bytecode info on a node to be created, without an originating bytecode node, I found it's the same as its first child's bytecode info (from trace logs). I haven't checked nor tested the bytecode info in cases where the first child is not available either, but I think the bytecode index can be arbitrary value because the bytecode info of the node is created using the default constructor, which doesn't initialize bytecode index nor `doNotProfile` flag.

There are too many places creating a node without passing an originating bytecode node after IL gen. Node API users are not aware of the need of originating bytecode node, there's no notes in the documentation nor assertion requiring such node. The fix you mentioned will need to touch basically every opt. It will also require knowing what the opt is doing and passing the right originating bytecode node, thus it is a large item. IMO, apart from fixes to where the Node API is used, the Node API itself should also be clear on what are needed. Maybe we should not allow all Node APIs to be usable after ILGen. In summary, I don't think it's this PR's job to fix that.

Reply:
..."but I think the bytecode index can be arbitrary value because the bytecode info of the node is created using the default constructor"...is a somewhat worrisome comment. 

I believe we did mandate that we pass in an originating bytecode node to node creation APIs in the past but it's possible some revision of that API in the last few years removed that mandatory requirement. I say this to suggest (hope ?) that there are'nt too many cases of creating nodes without using an API requiring an originating bytecode node but I agree it's probably better to deal with that in a separate PR. I would appreciate it if you would please create an issue for that and work on it asap once you have completed your current tasks because it's a dangerous state to be in to have uninitialized bytecode info on some nodes. On occasions the owning method is looked up based on the caller index that is part of the bytecode info and to have an uninitialized index to use to do the lookup could cause failures in some cases.

Reply:
I would appreciate @andrewcraik opinion on the above discussion as well.

Reply:
So the node creation/recreation APIs are a mess - there are a number of different versions with a number of different conventions and they do need to be cleaned up and rationalized, but I feel this is outside of the scope of this contribution discussion since @liqunl is proposing a perfectly reasonable pair of helpers as part of the OSR infrastructure.

The cases where the bytecode info won't be initialized - the cases of `createWithSymRef` which pass a NULL originating bytecode are especially problematic because the children are not available for the bytecode info. In general, bugs around the creation of nodes with uninitialized BCI have been seen a few times in the last year and indicate the API needs tightening up. That being said I don't think making the profiling flag handling explicit like has been done is wrong - it makes it clear what value it should have. I would argue that taking BCI info from a child isn't the most clear behavior either.

The issue with the node API is a much bigger one, but it will not be a quick fix. I think this is a good candidate for an architecture meeting and some help wanted items once we figure out which APIs are dangerous and need to be removed and what conventions we will apply to the handling of bytecodeinfo on recreate etc.

Reply:
@liqunl please open an issue for fixing cases when we have uninitialized bytecode info on nodes. 

@0xdaryl please consider this for discussion at the one of the upcoming OMR architecture meetings.

Reply:
https://github.com/eclipse/omr/issues/3153 is created for further discussion on bytecode info correctness.